src/sql/src/func.rs

```diff
@@ -3917,7 +3917,7 @@ pub static OP_IMPLS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(|| {
                     &ScalarType::String,
                 )?;
                 Ok(lhs.call_binary(rhs, TextConcat))
-            }) => String, 2779;
+            }) => NonVecAny, 2779;
```

Comment:
These return types were `String` before that PR, and these two lines were not changed there. Postgres confirms the expected output type is string:

```
mjibson=# select oid, oprresult::regtype from pg_operator where oid in (2779, 2780);
 oid  | oprresult 
------+-----------
 2779 | text
 2780 | text
```

Reply:
Then the type is wrong in Postgres too, at least for arrays:
```
deen=> select cast('{}' as text[]) || cast(null as text);
 ?column?
----------
 {NULL}
(1 row)
```
And for lists in Mz:
```
materialize=> select cast(list[] as text list) || cast(null as text);
 ?column?
----------
 {NULL}
(1 row)

materialize=> select pg_typeof(cast(list[] as text list) || cast(null as text));
 pg_typeof
-----------
 text list
(1 row)
```
But for other types it is correct as a string:
```
materialize=> select cast('x' as text) || cast(1 as int);
 ?column?
----------
 x1
(1 row)

materialize=> select pg_typeof(cast('x' as text) || cast(1 as int));
 pg_typeof
-----------
 text
(1 row)
```

I'll try to fix it.

Reply:
I'm pretty confused why this returns false for my calls with an empty list:
https://github.com/MaterializeInc/materialize/blob/main/src/repr/src/scalar.rs#L2316-L2321

Reply:
Might need to ask in slack. Sean knows how function resolution works pretty well and can probably explain.

Reply:
Ok, that's actually true, I'll check where it is going wrong then.